File: 1b - Functional Programming/sicp_chpt1.jl
# EXERCISES
# Exercise 1.8

function cube_root(x, guess=1, tolerance=0.001)
    close_enough(a) = abs(x - a^3) < tolerance
    improve_guess(a) = ((x / a^2) + 2a) / 3

    function newtons_method(y)
        if (close_enough(y))
            return y
        else
            newtons_method(improve_guess(y))
        end
    end

    return newtons_method(guess)
end

println(cube_root(27))


# Exercise 1.11
function f(n)
    if n < 3
        return n
    else
        return f(n - 1) + 2f(n - 2) + 3f(n - 3)
    end
end


println(f(14))


# prime test
function is_prime(n)
    for x in 2:sqrt(n)
        if n % x == 0
            return false
        end
    end
    return true
end

y = rand(2:123456789)
@time is_prime(y)
println("Is $y prime? $(is_prime(y))")

File: 1b - Functional Programming/sicp_chpt1.py
# EXERCISES

# Exercise 1.8: Newton's Method for cube root
def cube_root(x:float, guess:float = 1.0, tolerance:float = 0.001) -> float:
    def close_enough(a:float) -> bool: 
        return abs(x - (a * a * a)) < tolerance
    def improve_guess(a:float) -> float:
        return ((x / (a * a)) + 2 * a) / 3
    def newtons_method(a:float) -> float:
        if close_enough(a):
            return a
        else:
            return newtons_method(improve_guess(a))
    
    return newtons_method(guess)


print(cube_root(9))
print(cube_root(27))


# Exercise 1.11
def f(n:int) -> int: 
    if n < 3:
        return n
    else:
        return f(n - 1) + 2*f(n - 2) + 3*f(n - 3)

print(f(14))


# prime test
def is_prime(n:float) -> bool:
    for x in range(2, int(n.sqrt())):
        if n % x == 0:
            return False
    return True

import random
num = random.randint(1, 123456789)
print(is_prime(num))